beacon-chain/db/kv/validated_tips.go

```diff
@@ -0,0 +1,71 @@
+package kv
+
+import (
+	"context"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+)
+
+// ValidatedTips returns all the validated_tips that are present in the DB.
+func (s *Store) ValidatedTips(ctx context.Context) (map[[32]byte]types.Slot, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.ValidatedTips")
+	defer span.End()
+
+	valTips := make(map[[32]byte]types.Slot, 1)
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(validatedTips)
+
+		c := bkt.Cursor()
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+			valTips[bytesutil.ToBytes32(k)] = bytesutil.BytesToSlotBigEndian(v)
+		}
+		return nil
+	})
+	return valTips, err
+}
+
+// UpdateValidatedTips clears off all the old validated_tips from the DB and
+// adds the new tips that are provided.
+func (s *Store) UpdateValidatedTips(ctx context.Context, newVals map[[32]byte]types.Slot) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.UpdateValidatedTips")
+	defer span.End()
+
+	// Get the already existing tips.
+	oldVals, err := s.ValidatedTips(ctx)
+	if err != nil {
+		return err
+	}
+
+	updateErr := s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(validatedTips)
+
+		// Delete keys that are present and not in the new set.
+		for k, _ := range oldVals {
+			if _, ok := newVals[k]; !ok {
+				deleteErr := bkt.Delete(k[:])
```

Comment:
Perhaps I am missing something, I don't understand why deletion is necessary here. If the input data is in `key` `value` form, a simple `bkt.Put(key, value)` shall replace the previous `value`

Reply:
this is deleting the keys that are not in the updated version

Reply:
So this is all or nothing. Why go so far to store {`[32]byte` -> `types.Slot`} instead we could just do just use a fixed item key. {`[]byte("sync-tips-key")` -> `map[[32]byte]types.Slot`}. The latter seems easier to reason and better performance

Reply:
After more discussions, we concluded that the map would have a single entry most of the time, and we will be changing that entry. Keeping it this way is simpler